ARM: fix wrong stack bounds stored in thread

tyrantlizard77 opened the pull request:
Consider a stack buffer at address 0x10000 with size 1024.
If a thread is created with this stack object, the resulting
fields in thread.stack_info ended up being a base address of
0x10020 with size 1024. The guard size needed to be subtracted
from the size in order for the bounds to be correct.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
ARM: fix wrong stack bounds stored in thread

Consider a stack buffer at address 0x10000 with size 1024.
If a thread is created with this stack object, the resulting
fields in thread.stack_info ended up being a base address of
0x10020 with size 1024. The guard size needed to be subtracted
from the size in order for the bounds to be correct.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

ioannisg commented on arch/arm/core/thread.c:
```diff
@@ -91,10 +91,10 @@ void z_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
+
 	struct __esf *pInitCtx;
 
 	_new_thread_init(thread, pStackMem, stackSize, priority,
```
Hm, I think I know why this was missed; because initially I wanted to skip the workaround for power-of-2-alignment. The change was introduced here: a23a542b801235fc00bec7bae8d465e105f40f9a

Could you instead revert this and simply call _new_thread_init with stackEnd - stackMem?

andrewboie replied:
> Could you instead revert this and simply call _new_thread_init with stackEnd - stackMem?

Why is that preferable? stackSize is simply *wrong*

ioannisg replied:
I am still not sure this is a bug - let me take a closer look

andrewboie replied:
Oh it's a bug. You have, in the thread struct, a stack bounds setting which ends 32 bytes past the stack object!

andrewboie replied:
See https://github.com/zephyrproject-rtos/zephyr/pull/13650

ioannisg replied:
I am not convinced this is a bug; at least, not an active bug. The reason is simple; the `stack_info.size` needs to have the nominal size of the stack object, in order to put the whole area into the "user" MPU region. If you check the thread stack programming in `arm_core_mpu.c`, you will see that the MPU boundaries of the thread stack are `base = stack_obj `(not `stack_info.start`) and `stack_info.size`.

andrewboie replied:
It is a bug.

The stack info in the struct k_thread has the wrong boundaries. This information is supposed to be correct and it gets used in various places.

andrewboie replied:
```
starting test - test_stack_buffer
Reserved space: 0
Provided stack size: 1024
 - Stack object 0x20004800[1024]
 - Testing supervisor mode
   - Thread reports buffer 0x20004820 size 1024

    Assertion failed at /projects/zephyr2/zephyr/tests/kernel/mem_protect/userspace/src/main.c:982: stack_buffer_scenarios: (stack_end <= obj_end is false)
stack size in thread struct out of bounds (overflow)
FAIL - test_stack_buffer

```

ioannisg replied:
OK I see, I was focusing only on the MPU; there is clearly no bug there.

However, if this patch gets in, we need a fix in the MPU re-programming (on how to use the boundaries for the thread stack). I can take that over.

ioannisg commented on arch/arm/core/thread.c:
```diff
@@ -91,10 +91,10 @@ void z_new_thread(struct k_thread *thread, k_thread_stack_t *stack,
-	stackEnd = pStackMem + stackSize - MPU_GUARD_ALIGN_AND_SIZE;
-#else
-	stackEnd = pStackMem + stackSize;
+	stackSize -= MPU_GUARD_ALIGN_AND_SIZE;
```
Also, I am stressing this once more :) 
This workaround is only needed because we do not have a formal enforcement in the API of k_thread_create(), to use `size = K_THREAD_STACK_SIZEOF(K_THREAD_STACK_DEFINE(sym, <nominal size>))`

ioannisg committed:
arch: arm: no special case for MPUs with power-of-2-align requirement

Field stack_info.size is supposed to hold the actual writable
thread stack area, above what is pointed to by stack_info.start.
Therefore, in all MPU architectures we can program the Thread
Stack MPU region as starting from thread->stack_obj and with
size equal to stack_info.size plus the (possible) difference
between stack_obj and stack_info.start.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@agross-linaro @galak could you please, talk a quick look at this bug.

ioannisg commented:
@andrewboie could we get this in for 1.14? And, as early as possible cause we should be testing it in master for few days before 1.14

nashif commented:
do we have a bug we are tracking for this? or is this a PR=BUG=FIX?

ioannisg commented:
> do we have a bug we are tracking for this? or is this a PR=BUG=FIX?

It is a PR=BUG=FIX.

andrewboie commented:
> do we have a bug we are tracking for this? or is this a PR=BUG=FIX?

I didn't file a bug on this, but can if desired. It was something I discovered when beefing up the testing for thread stack characteristics.
